UWAZI - list templates

ValbuenaG opened the pull request:
For more information on contributing to Tella source code, see the [Contributor Guidelines](contributing/contributor_guide.md).

## Type of change

**Description:**


**Select the type of change(s) made in this pull request:**
- [ ] Bug fix *(Fixes an issue)*
- [x] New feature *(Adds functionality)*
- [ ] Documentation *(Fix to documentation)*

----------------------------------------------------------------------------------------

Fixes #issue-number


Loom update

https://www.loom.com/share/e4512da2e7394fe5957410b7393c00e1

ValbuenaG commented on Tella/Data/Database/TellaDataBase.swift:
```diff
@@ -675,5 +637,119 @@ extension TellaDataBase {
+        guard let locale = serversDict.first else { return nil }
+        return try self.parseDicToObjectOf(type: CollectedTemplate.self, dic: locale)
+    }
+    func getUwaziTemplateeWith(templateId: Int) throws -> CollectedTemplate? {
```
here is a typo on `getUwaziTemplateeWith` should be `getUwaziTemplateWith`

ValbuenaG commented on Tella/Data/Database/TellaDataBase.swift:
```diff
@@ -675,5 +637,119 @@ extension TellaDataBase {
+        statementBuilder.delete(tableName: D.tUwaziTemplate,
+                                primarykeyValue: [KeyValue(key: D.cTemplateId, value: templateId)])
+    }
+    func deleteAllUwaziTemplateWith(templateNo: Int) throws {
```
I think this naming is not quite descriptive. We are attempting to delete only one template but the method is called `deleteAllUwaziTemplateWith`

Also it would be more descriptive to  change `templateNo` to `id` or just `templateId`

ValbuenaG commented on Tella/Data/Database/TellaDataBase.swift:
```diff
@@ -675,5 +637,119 @@ extension TellaDataBase {
+        }
+        return []
+    }
+    func addUwaziTemplateWith(template: CollectedTemplate) throws -> CollectedTemplate {
```
why are we using the suffix "with" at the end of every method? shouldn't be more simple `addUwaziTemplate`?

rob-horizontal-iOS replied:
Ya that make scene I will remove it 
We used to add preposition to end of the function to make it more human readable.

ValbuenaG commented on Tella/Scenes/Uwazi/ViewModel/UwaziViewModel.swift:
```diff
@@ -0,0 +1,179 @@
+    func deleteDownloadedTemplate(templateId: Int) {
+        do {
+            _ = try self.mainAppModel.vaultManager.tellaData.deleteAllUwaziTemplateWith(templateNo: templateId)
+            downloadedTemplates = downloadedTemplates.filter({$0.id != templateId})
```
Here we could remove it by doing `downloadedTemplates.removeAll { $0.id == templateId }` instead of filtering and reassigning it, It is more direct that way.

dhekra-rouatbi commented on Tella/Data/Database/Protocols/UwaziServerLanguageProtocol.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+protocol UwaziServerLanguageProtocol {
+    func createLanguageTableForUwazi()
```
In my opinion it's better to create a separate Uwazi Server table and it contains the locale cookie and connect cookie  instead to use tella web server table and add language table 
see Android implementation:    `createUZIServer` https://github.com/Horizontal-org/Tella-Android/blob/b941e38f9d62032e1471b6dfd5e467b44acca23d/mobile/src/main/java/rs/readahead/washington/mobile/data/database/UwaziDataSource.java#L319C80-L319C80

dhekra-rouatbi commented:
I've added some remarks to this view 

![Simulator Screenshot - iPhone 15 Pro Max - 2023-09-22 at 09 08 59](https://github.com/Horizontal-org/Tella-iOS/assets/91288060/0dec22ce-da43-4538-a29d-ab2eb3eadc6f)

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/UwaziViewModel.swift:
```diff
@@ -0,0 +1,149 @@
+        Task {
+            guard let id = self.server.id else { return }
+            guard let locale = self.tellaData?.getUwaziLocale(serverId: id) else { return }
+            let template = try await UwaziServerRepository().handleTemplate(server: self.server, locale: locale)
```
I think it's better to include the locale to the server see https://github.com/Horizontal-org/Tella-iOS/pull/74/files#r1334325206

rob-horizontal-iOS commented on Tella/Data/Networking/Repositories/UwaziServerRepositories.swift:
```diff
@@ -78,6 +124,136 @@ struct UwaziServerRepository: WebRepository {
+    fileprivate func getAllowedTemplates(server: Server, settings: UwaziSetting, templates: [UwaziTemplateRow]?) -> [UwaziTemplateRow] {
+        // Check whether the server instance is public and if public then only use the whitelisted templates are added to resultTemplates
+        var tempTemplates: [UwaziTemplateRow] = []
+        if !self.isPublic(server: server) {
```
As per this [suggestion](https://github.com/Horizontal-org/Tella-iOS/pull/74#discussion_r1334325206) I think this is move to be a `property` like  `isPublic` of the new object example `UWaziUploadServer` like that is in android

ValbuenaG commented on Tella/Scenes/Uwazi/ViewModel/UwaziViewModel.swift:
```diff
@@ -0,0 +1,149 @@
+
+    @Published var templates : [CollectedTemplate] = []
+    @Published var downloadedTemplates : [CollectedTemplate] = []
+    @Published var draftReports : [Report] = []
```
Yes, you are right! I was supposed to change this once we start implementing the draft functionality but I can change that now

dhekra-rouatbi commented on Tella/Data/Networking/Repositories/UwaziServerRepositories.swift:
```diff
@@ -78,6 +124,136 @@ struct UwaziServerRepository: WebRepository {
+                let resultTemplates = self.getAllowedTemplates(server: server, settings: settings, templates: templates)
+                self.translate(locale: locale, resultTemplates: resultTemplates, translations: translations)
+                let originalTemplate = resultTemplates.map { template in
+                    return CollectedTemplate(serverId: serverID,
```
I think It's better to map the original template in the view model, (mabe later we will add an other layer that implements the call to the API and at the same time save to database)

dhekra-rouatbi commented on Tella/Data/Networking/Repositories/UwaziServerRepositories.swift:
```diff
@@ -56,19 +57,64 @@ struct UwaziServerRepository: WebRepository {
+    func getTemplate(serverURL: String, cookieList: [String]) -> AnyPublisher<UwaziTemplate, APIError> {
+        let apiResponse: APIResponse<UwaziTemplateDTO> = getAPIResponse(endpoint: API.getTemplate(serverURL: serverURL, cookieList: cookieList))
+        return apiResponse
+            .compactMap{$0.0.toDomain() as? UwaziTemplate}
```
I think in this level we don't need to convert the DTO to domain because we will use it later to handle the allowedPublicTemplates and translation ...

dhekra-rouatbi commented on Tella/Domain/Entity/Uwazi/Template/UwaziTemplateRow.swift:
```diff
@@ -0,0 +1,38 @@
+//
+
+import Foundation
+// TODO: Removed unwanted variables when done with entity flow no removed now due to confusion on what will be need in future implementation
```
In this case we need to remove the conversion `toDomain` from Uwazi template models, because in the current implementation the conversion is not implemented properly

ValbuenaG commented:
> I've added some remarks to this view
> 
> ![Simulator Screenshot - iPhone 15 Pro Max - 2023-09-22 at 09 08 59](https://user-images.githubusercontent.com/91288060/269942150-0dec22ce-da43-4538-a29d-ab2eb3eadc6f.png)

Just fixed this in this [commit](https://github.com/Horizontal-org/Tella-iOS/pull/74/commits/6e311beec6ba325a3e4a0045fab1fa36361b40cc) :)

ValbuenaG commented on Tella/Scenes/Settings/Models/ServerActionType.swift:
```diff
@@ -11,14 +11,37 @@ enum ServerActionType: ActionType {
     
     ListActionSheetItem(imageName: "edit-icon",
-                        content: "Edit",
+                        content: LocalizableUwazi.uwaziServerEdit.localized,
```
Thanks for the suggestion! In this case, this looks like a common sheet item, so I think creating a new ActionItem just for this would be an overkill, especially because we are not using any additional data. Only `imageName`, `content`, and `type`. Also, this `ListActionSheetItem` class is being used on all the codebase so creating a new one will be kind of inconsistent

dhekra-rouatbi replied:
Thanks, It's just for organisation and as I told you in the begining of the project I asked about sharing common string between views, I was answered that it's better to have each screen his own strings.

dhekra-rouatbi commented on Tella/Scenes/Uwazi/Common/TemplateItemView.swift:
```diff
@@ -0,0 +1,49 @@
+import SwiftUI
+
+struct TemplateItemView: View {
+    @Binding var template : CollectedTemplate
```
I suggest here to pass in parameter a `templateItemViewModel` that contains formatted parameters like template name, isDownloaded, downloadAction, deleteAction and when mapping from CollectedTemplate to  a templateItemViewModel  in the view model you can pass in parameter directly the action located in the viewmodel (downloadAction or deleteAction) so it will be cleaner and easier than to pass the whole `CollectedTemplate`

dhekra-rouatbi commented on Tella/Components/Reports/ReportsPageView.swift:
```diff
@@ -0,0 +1,62 @@
+
+import SwiftUI
+
+struct ReportsPageView: View {
```
sorry @ValbuenaG I didn't check this view carefully if it can be a shared component or not.
I don't think we can use `ReportsPageView`(ConnectionPageView) as a shared component because we have different type of views and parameters, so could you keep the old view in `ReportsView` and just fix the view in `UwaziView`

ValbuenaG replied:
yeah, you are right! I roll back to the previous view. My idea was to reuse the component in both and future connections but each connection has its own view model and specific stuff.

rob-horizontal-iOS commented on Tella/Data/Networking/Repositories/UwaziServerRepositories.swift:
```diff
@@ -78,6 +124,136 @@ struct UwaziServerRepository: WebRepository {
+
+    // Maps the options to the property of the template
+    fileprivate func handleMapping(_ templates: [UwaziTemplateRow]?, _ dictionary: [UwaziDictionaryRow]?) {
+        templates?.forEach { template in
```
For this we are just taking the value from the` dictionaryItem.values` and assigning to the `template.values` so compact is not necessary has it will return something after mapping and we do not need anything returned so. Both the compactMap and ForEach has same time complexity in this case i.e O(n) so there will not also be any performance bottleneck.

rob-horizontal-iOS commented on Tella/Scenes/Uwazi/ViewModel/UwaziViewModel.swift:
```diff
@@ -0,0 +1,149 @@
+        PageViewItem(title: LocalizableReport.submittedTitle.localized, page: .submitted, number: "")
+    ]
+    @Published var isLoading: Bool = false
+    @Published var serverURL : String
```
`serverName` is used in `UwaziView` for `serverURL` deleted it

ValbuenaG commented on Tella/Scenes/Uwazi/ViewModel/UwaziViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+
+
+class UwaziViewModel: ObservableObject {
+    
```
I created this `UwaziViewModel` to handle the general pageView logic of uwazi. That way we have the `UwaziTemplateViewModel` focusing only on template stuff. 

```
@Published var draftEntities : [Report] = []
    @Published var outboxedEntities : [Report] = []
    @Published var submittedEntities : [Report] = []
```

this should be type `UwaziEntity` but we have that type in another PR. I suggest we merge this with that type and then on the entity one I will change it with the correct one.

dhekra-rouatbi commented on Tella/Scenes/Uwazi/Views/Template/TemplateListView.swift:
```diff
@@ -0,0 +1,85 @@
+    
+    
+    //TODO: to move to view model
+    fileprivate func createCardViewModel(_ template: Binding<CollectedTemplate>) -> TemplateCardViewModel {
```
I suggest to move the mapping to the TemplateCardViewModel

dhekra-rouatbi commented on Tella/Components/Connections/ReportsPageView.swift:
```diff
@@ -0,0 +1,62 @@
+                                
+                    switch selectedCell {
+                    
+                    case .templates:
```
In Report view we don't have templates, I think it's better to create a separate enum for Reports and one for Uwazi

dhekra-rouatbi commented on Tella/Components/Tabs/Pages.swift:
```diff
@@ -6,13 +6,16 @@ import Foundation
     var title: String {
         switch self {
             
+        case .templates:
```
https://github.com/Horizontal-org/Tella-iOS/pull/74#discussion_r1338550736
If you can localise the strings for Uwazi and  Reports titles

dhekra-rouatbi commented on Tella/Scenes/Uwazi/Views/Template/TemplateListView.swift:
```diff
@@ -0,0 +1,85 @@
+    
+    private func showDeleteTemplateConfirmationView(template: CollectedTemplate) {
+        sheetManager.showBottomSheet(modalHeight: 200) {
+            let deleteViewModel = DeleteTemplateConfirmationViewModel(title: template.entityRow?.translatedName ?? "",
```
In my opinion it's unuseful to create DeleteTemplateConfirmationViewModel

rob-horizontal-iOS replied:
Removed DeleteTemplateConfirmationViewModel and DeleteTemplateConfirmationView and used ConfirmBottomSheet instead

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/TemplateCardViewModel.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+class TemplateCardViewModel {
+    var id = UUID()
```
What do you think if you use the template id (if exist) instead of creating a new id (UUID())

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/TemplateItemViewModel.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+
+class TemplateItemViewModel: Hashable {
+    var id = UUID()
```
https://github.com/Horizontal-org/Tella-iOS/pull/74#discussion_r1338633678

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/UwaziTemplateViewModel.swift:
```diff
@@ -0,0 +1,165 @@
+    var mainAppModel : MainAppModel
+    
+    @Published private var templates : [CollectedTemplate] = []
+    @Published private var downloadedTemplates : [CollectedTemplate] = []
```
In my opinion it's better to create a view model for Adding a template and other for getting downloaded Template(main view), What do you think?

ValbuenaG replied:
I created a new `AddTemplateVM` and `DownloadedTemplatesVM` so we divide the responsibilities between both view models

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/DownloadedTemplatesVM.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    @Published var templateCardsViewModel : [TemplateCardViewModel] = []
+    
+    @Published var isLoading: Bool = false
```
I think isLoading, serverName and server are unused

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/DownloadedTemplatesVM.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    /// To determine if the templates are already download or not reflect on the UI for template download list
+    /// - Parameter templates: Collection of CollectedTemplate to determine if it downloaded or not
+    func handleTemplateDownload(templates: [CollectedTemplate]) {
```
I think handleTemplateDownload is unused

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/DownloadedTemplatesVM.swift:
```diff
@@ -0,0 +1,75 @@
+
+    
+    func getDownloadedTemplates() {
+        self.downloadedTemplates = self.getAllDownloadedTemplate() ?? []
```
What do you think if you update this to 
` self.downloadedTemplates =  self.tellaData?.getAllUwaziTemplate()` and remove  `getAllDownloadedTemplate` function

dhekra-rouatbi commented on Tella/Scenes/Uwazi/ViewModel/UwaziTemplateViewModel.swift:
```diff
@@ -0,0 +1,165 @@
+    
+    func downloadTemplate(template: CollectedTemplate) {
+        var template = template
+        Toast.displayToast(message: "“\(template.entityRow?.translatedName ?? "")” “\(LocalizableUwazi.uwaziAddTemplateSavedToast.localized)”")
```
I think Toast is a view and it's better to display it in the view not in the view model
